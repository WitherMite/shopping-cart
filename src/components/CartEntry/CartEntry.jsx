import PropTypes from "prop-types";

const USDFormatter = Intl.NumberFormat("en-US", {
  style: "currency",
  currency: "USD",
});

CartEntry.propTypes = {
  onIncrement: PropTypes.func,
  onDecrement: PropTypes.func,
  onDelete: PropTypes.func,
  image: PropTypes.string,
  title: PropTypes.string.isRequired,
  price: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  count: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]),
};

export default function CartEntry({
  image,
  title,
  price,
  count,
  onIncrement,
  onDecrement,
  onDelete,
  children,
}) {
  return (
    <section className="cart-item-card">
      <img src={image} alt="" className="item-img" />
      <h2 className="item-title">{title}</h2>
      {children}
      <p className="item-price">{USDFormatter.format(price) + " / item"}</p>
      <div className="item-count-field">
        <button onClick={onDecrement} className="cart-subtract-item-btn">
          -
        </button>
        <p className="item-count">{count}</p>
        <button onClick={onIncrement} className="cart-add-item-btn">
          +
        </button>
      </div>
      <p className="entry-price">{USDFormatter.format(price * count)}</p>
      <button onClick={onDelete} className="cart-remove-item-btn">
        Remove from Cart
      </button>
    </section>
  );
}
