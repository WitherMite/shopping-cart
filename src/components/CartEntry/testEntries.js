import logo from "../../assets/logo-ipsum-small.svg";

const testEntries = [
  {
    id: 1,
    title: "Item 1",
    price: "40",
    category: "test",
    description:
      "This is item number 1, there are many like it, but this is the first",
    image: logo,
    count: 1,
  },
  {
    id: 2,
    title: "Item 2",
    price: "12",
    category: "test",
    description: "Lorem ipsum sit amet dolor...",
    image: logo,
    count: 50,
  },
  {
    id: 3,
    title: "Item 3",
    price: "20.45",
    category: "test",
    description: "Three.",
    image: logo,
    count: 3,
  },
];

export default testEntries;
