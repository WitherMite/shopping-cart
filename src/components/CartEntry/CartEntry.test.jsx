import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import testEntries from "./testEntries.js";
import CartEntry from "./CartEntry.jsx";

describe("CartEntry component", () => {
  it("Renders entry info and component's children", () => {
    const childEle = <div>Hello there.</div>;
    const { container } = render(
      <CartEntry {...testEntries[0]}>{childEle}</CartEntry>
    );

    expect(container).toMatchSnapshot();
  });

  it("Clicking + button calls onIncrement prop", async () => {
    const user = userEvent.setup();
    const fn = vi.fn();
    render(<CartEntry {...testEntries[0]} onIncrement={fn} />);

    const button = screen.getByRole("button", { name: "+" });
    await user.click(button);

    expect(fn).toHaveBeenCalled();
  });

  it("Clicking - button calls onDecrement prop", async () => {
    const user = userEvent.setup();
    const fn = vi.fn();
    render(<CartEntry {...testEntries[1]} onDecrement={fn} />);

    const button = screen.getByRole("button", { name: "-" });
    await user.click(button);

    expect(fn).toHaveBeenCalled();
  });

  it("Clicking Remove from Cart button calls onDelete prop", async () => {
    const user = userEvent.setup();
    const fn = vi.fn();
    render(<CartEntry {...testEntries[2]} onDelete={fn} />);

    const button = screen.getByRole("button", { name: "Remove from Cart" });
    await user.click(button);

    expect(fn).toHaveBeenCalled();
  });
});
