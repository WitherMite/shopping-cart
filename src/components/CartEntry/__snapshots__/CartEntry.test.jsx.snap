// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartEntry component > Renders entry info and component's children 1`] = `
<div>
  <section
    class="cart-item-card"
  >
    <img
      alt=""
      class="item-img"
      src="/src/assets/logo-ipsum-small.svg"
    />
    <h2
      class="item-title"
    >
      Item 1
    </h2>
    <div>
      Hello there.
    </div>
    <p
      class="item-price"
    >
      $40.00 / item
    </p>
    <div
      class="item-count-field"
    >
      <button
        class="cart-subtract-item-btn"
      >
        -
      </button>
      <p
        class="item-count"
      >
        1
      </p>
      <button
        class="cart-add-item-btn"
      >
        +
      </button>
    </div>
    <p
      class="entry-price"
    >
      $40.00
    </p>
    <button
      class="cart-remove-item-btn"
    >
      Remove from Cart
    </button>
  </section>
</div>
`;
